src/compiler/scanner.ts

```diff
@@ -858,28 +858,76 @@ namespace ts {
             }
         }
 
+        function scanNumberFragment(): string {
+            let start = pos;
+            let allowSeperator = false;
+            let result: string;
+            while (true) {
+                const ch = text.charCodeAt(pos);
+                if (allowSeperator && ch === CharacterCodes._) {
+                    allowSeperator = false;
+                    tokenFlags |= TokenFlags.ContainsSeparator;
+                    result = (result || "") + text.substring(start, pos);
+                    pos++;
+                    start = pos;
+                    continue;
+                }
+                if (isDigit(ch)) {
+                    allowSeperator = true;
+                    pos++;
+                    continue;
+                }
+                break;
+            }
+            if (text.charCodeAt(pos - 1) === CharacterCodes._) {
+                pos--;
+                error(Diagnostics.Numeric_seperators_are_not_allowed_at_the_end_of_a_literal, 1);
+                pos++;
+            }
+            return result = (result || "") + text.substring(start, pos);

```

Comment:
You definitely assign `result` to at least `""`, so why not just initialize `result` above?

Reply:
Also, you don't need to reassign `result` here.

Reply:
> You definitely assign `result` to at least `""`, so why not just initialize `result` above?

`result` is only assigned in the presence of a separator.

> Also, you don't need to reassign `result` here.

Done.